fix build errors with nr53 net core and power management

pabigot opened the pull request:
The net core of nRF53 doesn't provide access to the power regulator features: they can be accessed only from the application core.

Revert the previous way of attempting to avoid the test problem, and exclude the net cores from the test.

Fixes #31627

pabigot committed:
Revert "soc: nordic: Disable PM for SOC_NRF5340_CPUNET_QKAA"

This reverts commit f38ba314ea29a34931453e8eed59e27bd4ae4da4.

nrf5340dk_nrf5340_cpunet doesn't support power management, but that
should be handled another way.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
tests: exclude nrf5340dk_nrf5340_cpunet when CONFIG_PM=y

The required regulator control features are only available on the
application core.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

ceolin commented:
You are just excluding this target from the test. But one still can enable it and the build will fail. This target does not support power management.

ceolin commented:
+1 for the commit excluding the target from the test. But this board still should not select PM

pabigot commented:
@ceolin Added a patch that disables PM on the net boards.  OK?

nashif commented:
> You are just excluding this target from the test. But one still can enable it and the build will fail. This target does not support power management.

i can also try and build a bluetooth sample/test for something that does not have bluetooth and it will blow up, we cant do the filtering on the subsystem level.

ceolin commented:
The problem in reverting that patch is that pm will be enabled by default for that target. That is different from one intentionally enable it.

ceolin commented:
> @ceolin Added a patch that disables PM on the net boards. OK?

that kinda works because this evaluated later. I'm fine with this solution for now. The solution having the hidden option seems more correct because the error will be spotted during the configuration though.

nashif commented:
> The problem in reverting that patch is that pm will be enabled by default for that target.

enabled by default how? where?

ceolin commented:
> > The problem in reverting that patch is that pm will be enabled by default for that target.
> 
> enabled by default how? where?

The commit that he is reverting has:
```
config PM
	default y if SYS_CLOCK_EXISTS
```

So PM will be enabled even for a hello world in that platform.

nashif commented:
> The commit that he is reverting has:

yes, I do not think this is right and not sure how is this part of the fix TBH.

pabigot commented:
> > The commit that he is reverting has:
> 
> yes, I do not think this is right and not sure how is this part of the fix TBH.

I'm confused.  The original patch f38ba314ea29a34931453e8eed59e27bd4ae4da4 from #31273 not only added the condition specific to `SOC_NRF5340_CPUNET_QKAA`, it also removed the default-enable of PM on all Nordic platforms.  Maybe that's a right thing to do, but it shouldn't have been done as a side effect of trying to fix the target-specific build problem, and I would have blocked the PR had I noticed it among everything else.

I agree #31627 is more general solution but unless we discuss it in Nordic-space I want the default for `CONFIG_PM` to remain what it has been since 5ec96b176689e192d205d4952df8ef23b17f1511 in May 2019, when it was automatically enabled.

pabigot commented:
Updated to exclude the cpunet board on all tests that explicitly enable `CONFIG_PM`.  Maybe this time it'll pass.

ceolin commented:
> > > The commit that he is reverting has:
> > 
> > 
> > yes, I do not think this is right and not sure how is this part of the fix TBH.
> 
> I'm confused. The original patch [f38ba31](https://github.com/zephyrproject-rtos/zephyr/commit/f38ba314ea29a34931453e8eed59e27bd4ae4da4) from #31273 not only added the condition specific to `SOC_NRF5340_CPUNET_QKAA`, it also removed the default-enable of PM on all Nordic platforms. Maybe that's a right thing to do, but it shouldn't have been done as a side effect of trying to fix the target-specific build problem, and I would have blocked the PR had I noticed it among everything else.
> 
> I agree #31627 is more general solution but unless we discuss it in Nordic-space I want the default for `CONFIG_PM` to remain what it has been since [5ec96b1](https://github.com/zephyrproject-rtos/zephyr/commit/5ec96b176689e192d205d4952df8ef23b17f1511) in May 2019, when it was automatically enabled.

It was just working because the old implementation had other conditionals that were avoiding the code to be compiled. In other words, this specific target was building with PM support but the implementation was empty.
To me is completely fair decide if your targets will have PM enabled by default, but in this case we need to come up with a solution that exclude this one.

ceolin reviewed (approved):
I still think we need to discuss this case in https://github.com/zephyrproject-rtos/zephyr/pull/31722
But this patch fixes the problem and restore the old behavior.

ioannisg commented on soc/arm/nordic_nrf/Kconfig.defconfig:
```diff
@@ -27,6 +27,9 @@ config SYS_CLOCK_TICKS_PER_SEC
 config ARCH_HAS_CUSTOM_BUSY_WAIT
 	default y
 
+config PM
```
I don't think the right place for a board option is the soc directory, just because of the board-soc hierharhy.

pabigot replied:
This is restoring behavior that existed for well over a year until it was removed as a side effect.  Are you sure you want to block on this?

ioannisg commented on boards/arm/nrf5340dk_nrf5340/nrf5340dk_nrf5340_cpunet_defconfig:
```diff
@@ -4,6 +4,9 @@ CONFIG_SOC_SERIES_NRF53X=y
 CONFIG_BOARD_NRF5340DK_NRF5340_CPUNET=y
 
+# Disable PM, not supported on net core
+CONFIG_PM=n
```
Could agree on the patch, if there could at least be a way to achieve the desired behavior, while not having to undefine something at board_defconfig

pabigot replied:
Can you recommend one?

ioannisg replied:
Yes. You could have a Kconfig capability, e.g. config CPU_SUPPORTS_PM, and add a dependency, so you won't need to set to =n here.

pabigot replied:
@ioannisg That'd be #31722 which is stalled because it's not a great solution either, since virtually every board would have to set it *and* it isn't clear what it should mean.  See https://github.com/zephyrproject-rtos/zephyr/pull/31722#discussion_r565719653

Seems to me it's worth getting in something that closes the high priority issue, then planning a more general solution later when we make more progress on PM.  OTOH I can just remove this, so the tests pass but anybody attempting to build samples like `hello_world` on `cpunet` will encounter a compile-time error.  Or we can keep the unintended removal of default-enabling of `CONFIG_PM` on all Nordic platforms that's been in place since 2019, but that might break out-of-tree applications. 

I think this solution's good enough for now.

pabigot committed:
power: allow finer control of PM selection

Some targets, such as network cores on a multi-core SoC, do not
support system power management because they do not control the SoC.
Provide a flag that can be selected to inihibit enabling system power
management on those targets.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
soc: nordic: disable system power management on nRF53 network core

The network core doesn't have access to the regulators necessary to
support system power management.  That has to be done on the
application core.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot commented:
Maybe this will come closer to satisfying everybody:

I took @ceolin's idea from #31722 but inverted the logic, so we have `HAS_NO_SYS_PM` that can be selected on targets that will never be responsible for system power management.  This gets rid of `CONFIG_PM=n` defaults on nRF53 network core.

Nordic continues to default-select PM, as [it has for almost two years](https://github.com/zephyrproject-rtos/zephyr/pull/31714#issuecomment-769017058.  I understand there are arguments that this shouldn't be done (@nashif), or should be done another way (@ioannisg) but this is a bug-fix, not a vehicle to change long-standing policy in Nordic devices or refactor Nordic default selection.

If default-enabling PM on Nordic platforms continues to block progress on fixing the nightly builds *which have been broken for over a week now*, then I would like @carlescufi to approve the change in behavor and I'll remove that default from `soc/arm/nordic_nrf/Kconfig.defconfig`

ioannisg commented:
Last version satisfies my suggestion for not forcing PM=n in the defconfig and map this to a SoC capability-based dependence:

> Could agree on the patch, if there could at least be a way to achieve the desired behavior, while not having to undefine something at board_defconfig

Approving this for the reasons @pabigot mentions; we should prioritize bug fixes for now and re-think how to improve this later.

nashif commented:
we need to revisit this change and some of the things that were added,  The policy for not enabling PM by default should be a zephyr policy and not per vendor. PM should be only enabled when you are testing PM, nothing else.
